Shell handler returns int value

jakub-uC opened the pull request:
Shell handler can now return an int value. This gives an opportunity to check if command has been executed successfully. It will allow to write tests for modules using shell commands. 

This change is needed to make new shell backward compatible with legacy shell.

Commit:
subsys: shell: add int ret_val to command handlers

1. Command handler can return command exectution status as int.
2. Existing command handlers rework.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

Commit:
doc: subsystems: shell: update documentation

Shell handler can now return an int value. This change has
been reflected in shell.rst.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

nordic-krch commented on subsys/shell/shell.c:
```diff
@@ -765,7 +765,7 @@ static void shell_state_collect(const struct shell *shell)
 				} else {
 					/* Command execution */
-					shell_execute(shell);
+					(void)shell_execute(shell);
```
you could add at least log message here if shell execution fails.

jarz-nordic replied:
I would like to avoid that, in my opinion logging failed execution should be implemented in command handler.

nordic-krch replied:
ok

nordic-krch commented on subsys/shell/shell_cmds.c:
```diff
@@ -177,127 +177,137 @@ static int terminal_size_get(const struct shell *shell)
+static int cmd_bacskpace_mode_backspace(const struct shell *shell, size_t argc,
 					 char **argv)
 {
 	if (!shell_cmd_precheck(shell, (argc == 1), NULL, 0)) {
```
this is not that simple: precheck may also fail if help was requested. In that case return 0 is ok but if it fails because of number of arguments then it should return error.

jarz-nordic replied:
Yes, this is good point. It will be fixed in other PR.

Commit:
subsys: shell: add handlers diagnostic function

Added function: shell_execute_cmd that can be called
for command diagnostic purposes.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

nvlsianpu commented:
IMO changes of the looker shell should goes in dedicated commit.

carlescufi commented:
@rakons you might want to wait until this is merged for you rebase

carlescufi commented on include/shell/shell.h:
```diff
@@ -71,11 +71,13 @@ struct shell_cmd_entry {
 
 struct shell;
 
+#define SHELL_CMD_EXECUTED_ERROR	((int)-2)
```
Please use `errno.h` types instead

carlescufi replied:
https://github.com/zephyrproject-rtos/zephyr/blob/master/lib/libc/minimal/include/errno.h#L32

carlescufi replied:
Maybe `return -ENOEXEC`

jarz-nordic replied:
done

Vudentz commented on samples/subsys/shell/shell_module/src/main.c:
```diff
@@ -29,41 +29,46 @@ void timer_expired_handler(struct k_timer *timer)
+static int cmd_log_test_start(const struct shell *shell, size_t argc,
+			      char **argv, u32_t period)
 {
 	if (!shell_cmd_precheck(shell, argc == 1, NULL, 0)) {
```
Shouldn't we go one step further and replace the calls to shell_cmd_precheck on the handlers with just a specific error so the shell would take care of calling shell_cmd_precheck internally?

carlescufi commented:
@Vudentz @nordic-krch this should be merged soon since it will impact the transition of other subsystems to the new shell.

Vudentz reviewed (approved):
LGTM
